Gibbs removal

RafaelNH opened the pull request:
In this pull request, I am submitting a new pre-processing algorithm to remove Gibbs Ringing artefacts.

This algorithm, is an adaption of the algorithm suggested by: 

Kellner et al. Gibbs‐ringing artifact removal based on local subvoxel‐shifts. MRM 2016

The full adapted version of algorithm is described in chapter 3 of my phd thesis:

[Neto Henriques, R. (2018). Advanced Methods for Diffusion MRI Data Analysis and their Application to the Healthy Ageing Brain (Doctoral thesis). https://doi.org/10.17863/CAM.29356](https://www.repository.cam.ac.uk/handle/1810/281993)

pep8speaks commented:
Hello @RafaelNH, Thank you for submitting the Pull Request !

* In the file [`dipy/denoise/gibbs.py`](https://github.com/nipy/dipy/blob/1eb0b156160bd1e55bfa47f74c326914f3f92e4e/dipy/denoise/gibbs.py):

> [Line 253:19](https://github.com/nipy/dipy/blob/1eb0b156160bd1e55bfa47f74c326914f3f92e4e/dipy/denoise/gibbs.py#L253): [E711](https://duckduckgo.com/?q=pep8%20E711) comparison to None should be 'if cond is None:'

Do see the [DIPY coding Style guideline](https://github.com/nipy/dipy/blob/master/doc/devel/coding_style_guideline.rst)

arokem commented on dipy/denoise/gibbs.py:
```diff
@@ -0,0 +1,310 @@
+        xs = x.T.copy()
+
+    # TV for shift zero (baseline)
+    TVR, TVL = image_tv(xs, a=1, nn=nn)
```
Should the axis variable `a` really always be set to `1`? Can't you use the input to avoid the if statement in L88?

arokem commented on dipy/denoise/gibbs.py:
```diff
@@ -0,0 +1,310 @@
+        return xs.T
+
+
+def gibbs_weigthing_functions(shape):
```
How about renaming this `gibbs_weighting`? Or even just `weighting`? It is already in the `gibbs` module.

arokem commented on dipy/denoise/gibbs.py:
```diff
@@ -0,0 +1,310 @@
+        set to 3.
+    G0 : 2D ndarray, optional.
+        Weights for the image corrected along axis 1. If not given, the
+        function estimates them using function:
```
Use the sphinx directive: 

```
:func:`gibbs_weighting_functions`

```

arokem commented on dipy/denoise/tests/test_gibbs.py:
```diff
@@ -0,0 +1,135 @@
+image_gt[4 * Nre: 5 * Nre, 3 * Nre: 5 * Nre] = 3
+
+# Suppressing gibbs artefacts
+image_cor = gibbs_removal_2d(image_gibbs)
```
I think that it's best to put this call inside your test functions.

arokem commented on dipy/denoise/gibbs.py:
```diff
@@ -0,0 +1,310 @@
+    return imagec
+
+
+def gibbs_removal(vol, slice_axis=2, nn=3):
```
So this will be the main public interface for this module? This is the function that most users will want to access? I find the naming scheme here a little bit confusing. Maybe the others (the "sub-functions") should be designated as private functions by adding an underscore prefix?

arokem commented on doc/examples/reconst_dki.py:
```diff
@@ -87,7 +87,7 @@
 
 """
-Function ``read_cenir_multib`` return img and gtab which contains respectively
+Function ``read_cfin_multib`` return img and gtab which contains respectively
```
Good catch!

arokem commented on doc/examples/suppress_gibbs.py:
```diff
@@ -0,0 +1,155 @@
+[1]_, [2]_.
+
+In the following example, we show how to suppress Gibbs artefacts of MRI images
+in dipy. This algorithm is based on an adapted version of the sub-voxel
```
You don't need "in dipy" here.

arokem commented on doc/examples/suppress_gibbs.py:
```diff
@@ -0,0 +1,155 @@
+In the following example, we show how to suppress Gibbs artefacts of MRI images
+in dipy. This algorithm is based on an adapted version of the sub-voxel
+Gibbs suppression procedure [3]_. Full details of the implemented algorithm
+can be found in the Chapter 3 of [4]_  (please cite [3]_, [4]_ if you are using
```
"the Chapter" => "chapter"

arokem commented on doc/examples/suppress_gibbs.py:
```diff
@@ -0,0 +1,155 @@
+# -*- coding: utf-8 -*-
```
Maybe rename this file "denoise_gibbs.py"?

RafaelNH commented:
@arokem! Thanks for your review. I am still not sure if this should go under the denoising heather because I don't see Gibbs suppression as a denoising algorithm. But, anyway, I also agree that perhaps is not efficient to create a new heather just for this algorithm. Saying this, I think I've done all that you asked me on your review. Let's me know if I skipped something by distraction or if you have any further comments. Thanks ;), Rafael

RafaelNH commented:
@arokem - just realised that I didn't  address the comment about avoiding the if statement in line 88. Bit busy at the moment but I will do it asap!

skoudoro commented on dipy/denoise/tests/test_gibbs.py:
```diff
@@ -0,0 +1,151 @@
+                                gibbs_removal, _image_tv)
+from numpy.testing import (assert_, assert_array_almost_equal, assert_raises)
+
+# Produce a 2D image
```
Can you move this part to a `setup_module()` function like [here](https://github.com/nipy/dipy/blob/master/dipy/reconst/tests/test_fwdti.py#L23). Do not forget to explicit what is global or not

RafaelNH replied:
Very nice suggestion!

skoudoro commented on dipy/denoise/tests/test_gibbs.py:
```diff
@@ -0,0 +1,151 @@
+    assert_array_almost_equal(image4d_cor[:, :, 1, 1], image_cor)
+
+
+def test_swaped_gibbs_3d():
```
Can you add `test_swaped_gibbs_2d()` ?

RafaelNH replied:
Done!

skoudoro commented on dipy/denoise/gibbs.py:
```diff
@@ -0,0 +1,314 @@
+    ----------
+    vol : array ([X, Y, Z]) or ([X, Y, Z, g])
+        Matrix containing one volume (3D) or multiple (4D) volumes of images.
+    slice_axis : int (0, 1, or 2)
```
To be consistent with our framework, I think the user should format its data before calling your function.

RafaelNH replied:
Indeed, I expect that the data is ordered in the following way ([X, Y, Z, g]); however that does not tell which axis contains the different 2D acquisitions. Normally, the last axis is used for different slices acquisition (i.e. you acquire slices with coordinates X, Y and align then in Z), but not sure how this is general for all datasets (e.g. I already saw a dataset in which different slices were aligned in X coordinates).

RafaelNH replied:
Just to give a concrete example - imagine that you've acquired 60 slices with sampling matrix of 96x96. If you align image slices along the z axis, you will have a data matrix with shape (96, 96, 60). However, if user aligns the slices along the x-axis, your data matrix will have a shape of (60, 96, 96).  To use this code you will need to inform in which axis the slices are oriented because Gibbs suppression is done slice by slice.

skoudoro replied:
I understand your point but this is not coherent with all DIPY framework. the user will have to do the whole pipeline: denoise, unrig,  DTI  then tracking for example. So in your case, when data are flipped, it will work for your function but not the rest of the pipeline with strange to have 2 different logic. Give this opportunity to swap axis is misleading.  

It think it would be better for the user to flip the data at the beginning of the pipeline and then apply it to the whole process. 

Other opinions are welcomed

RafaelNH replied:
@skoudoro - this is an optional parameter. If user don’t call it, function will assume that data is already aligned. So you can still used it if user flips the data before at the beginning of pipeline.

Also it is important to note that even if user decides to use this option to flip the data, Gibbs_removal will still return the data in its original shape. So I don’t see why it is not coherent with Dipy. 

Personally I don’t have any issue in removing this swap option. However, I really don’t see the harm of having this function with this feature. I think function in this way is more general.

RafaelNH replied:
@skoudoro - this is an optional parameter. If user don’t call it, function will assume that data is already aligned. So you can still used it if user flips the data before at the beginning of pipeline.

Also it is important to note that even if user decides to use this option to flip the data, Gibbs_removal will still return the data in its original shape. So I don’t see why it is not coherent with Dipy. 

Personally I don’t have any issue in removing this swap option. However, I really don’t see the harm of having this function with this feature. I think function in this way is more general.

RafaelNH replied:
Actually it is very dangerous to flip the data at the beginning of the pipeline. Because if user does it, we have to make sure that the affine matrix and gradient tables are also properly flipped. That is way, flipping and unflipping data should be done internally in this function.

RafaelNH replied:
Actually it is very dangerous to flip the data at the beginning of the pipeline. Because if user does it, we have to make sure that the affine matrix and gradient tables are also properly flipped. That is way, flipping and unflipping data should be done internally in this function.

skoudoro replied:
> I expect that the data is ordered in the following way ([X, Y, Z, g]); however that does not tell which axis contains the different 2D acquisitions. Normally, the last axis is used for different slices acquisition (i.e. you acquire slices with coordinates X, Y and align then in Z), but not sure how this is general for all datasets (e.g. I already saw a dataset in which different slices were aligned in X coordinates)

ok, got it. you can let this option as it is. Thanks for the clarification!

RafaelNH commented on dipy/denoise/gibbs.py:
```diff
@@ -0,0 +1,314 @@
+        raise ValueError("Different slices have to be organized along" +
+                         "one of the 3 first matrix dimensions")
+    elif slice_axis < 2:
+        vol = np.swapaxes(vol, slice_axis, 2)
```
Data is already in the format assumed by the function when vol.ndim == 2, therefore you don't need that case in the elif statements. In the tests, I am testing a 2D image and all seems to run fine. I will add in the function documentation that data can also be a ([X, Y]) matrix.

skoudoro replied:
In your test, you do not swap axis with the 2d data and assume that the user will do the correct move :smile: 

if you do the same test with `swap_axis=0` (which is possible), this should crash. Since you are managing 2D data as well, it will be good to manage this case.

RafaelNH replied:
Very well spotted! I've just fixed the issue! For 2D data, swap_axis is a dummy variable, because for 2D data you need just to run _gibbs_removal_2d. See more information in the comments of my code.

arokem replied:
Is this resolved? If yes, please hit the "resolve conversation" button

RafaelNH commented on dipy/denoise/gibbs.py:
```diff
@@ -0,0 +1,314 @@
+
+    Parameters
+    ----------
+    vol : array ([X, Y, Z]) or ([X, Y, Z, g])
```
Indeed,...

RafaelNH commented on dipy/denoise/gibbs.py:
```diff
@@ -0,0 +1,314 @@
+
+    # Run gibbs removal of 2D images
+    if nd == 2:
+        vol = _gibbs_removal_2d(vol, n_points=n_points, G0=G0, G1=G1)
```
True! However, to do this change I will need to add two extra if statements, one in "check matrix dimension" and another in "Reshape data to original format". Since my code is not crashing for 2D image I will leave the code as it is.

RafaelNH commented:
@arokem - I've just recalled why I am using the if statement in line 88 instead of using the axis information. Basically, with this if statement, indexing in lines 121-135 is easier to read.

@skoudore - I've addressed most of your comments - I will do the rest asap (2 comments are missing)

RafaelNH commented:
@arokem , @skoudoro - I think I've addressed all your comments - let me know if you spot something else that I have to address. Many thanks for your inputs!

Borda commented on dipy/denoise/gibbs.py:
```diff
@@ -0,0 +1,319 @@
+    computed using the time series values on the right boundary and ptv values
+    are computed using the time series values on the left boundary.
+    """
+    if axis:
```
what about `xs = x.copy() if axis else x.T.copy()`

Borda commented on dipy/denoise/gibbs.py:
```diff
@@ -0,0 +1,319 @@
+    # original grid points
+    xs[idx] = (isp[idx] - isn[idx])/(sp[idx] + sn[idx])*sn[idx] + isn[idx]
+
+    if axis:
```
what about `return xs if axis else xs.T`

Garyfallidis commented on dipy/denoise/gibbs.py:
```diff
@@ -0,0 +1,319 @@
+
+def _weights(shape):
+    """ Computes the weights necessary to combine two images processed by
+    the 1D gibbs removal procedure along two different axis [1]_.
```
axes

Garyfallidis commented on dipy/denoise/gibbs.py:
```diff
@@ -0,0 +1,319 @@
+    K1, K0 = np.meshgrid(k[1:-1], k[1:-1])
+    cosk0 = 1.0 + np.cos(K0)
+    cosk1 = 1.0 + np.cos(K1)
+    G1[1:-1, 1:-1] = cosk0 / (cosk0+cosk1)
```
``cosk0 + cosk1``

Garyfallidis commented on dipy/denoise/gibbs.py:
```diff
@@ -0,0 +1,319 @@
+    cosk1 = 1.0 + np.cos(K1)
+    G1[1:-1, 1:-1] = cosk0 / (cosk0+cosk1)
+    G0[1:-1, 1:-1] = cosk1 / (cosk0+cosk1)
+
```
same

Garyfallidis commented on dipy/denoise/gibbs.py:
```diff
@@ -0,0 +1,319 @@
+    # check matrix dimension
+    if nd == 4:
+        inishap = vol.shape
+        vol = vol.reshape((inishap[0], inishap[1], inishap[2]*inishap[3]))
```
pep8

Garyfallidis commented on dipy/denoise/gibbs.py:
```diff
@@ -0,0 +1,319 @@
+    ----
+    This function was created to deal with gibbs artefacts of MR images.
+    Assuming that MR images are reconstructed from estimates of their Fourier
+    expansion coefficients, during TV calculation matrix x can taken as and
```
as and?

Garyfallidis commented on doc/examples/denoise_gibbs.py:
```diff
@@ -0,0 +1,155 @@
+ax.flat[1].set_title('Corrected')
+
+plt.show()
+fig1.savefig('Gibbs_suppression_b0.png')
```
Very hard to see the Gibbs artifact in the original data.

Garyfallidis commented on dipy/denoise/gibbs.py:
```diff
@@ -0,0 +1,319 @@
+    Note
+    ----
+    This function decreases the effects of gibbs oscillations based on the
+    analysis of local total variation (TV) along the two axis of the image.
```
two axes

Garyfallidis commented on dipy/denoise/gibbs.py:
```diff
@@ -0,0 +1,319 @@
+    Although artefact correction is done based on each point primary adjanced
+    neighbors, total variation should be accessed in a larger range of
+    neigbors. If you want to adjust the number of the neigbors to be
+    considered in TV calculation please change parameter nn.
```
Where is that nn parameter?

RafaelNH commented:
@arokem , @Garyfallidis , @skoudoro, @Borda - FYI: I am done with this PR. Continuous integration issues are not related to this PR. Please let me know of there is something else that you want me to address here.

skoudoro commented:
Overall, looks good to me, I just need to play a bit more with the example but I am ok to merge it by the end of this week if there is no other comment

arokem commented on dipy/denoise/gibbs.py:
```diff
@@ -0,0 +1,300 @@
+    ----
+    This function suppresses the effects of Gibbs oscillations based on the
+    analysis of local total variation (TV). Although artefact correction is
+    done based on two adjanced points for each voxel, total variation should be
```
```suggestion
    done based on two adjacent points for each voxel, total variation should be
```

arokem commented on dipy/denoise/gibbs.py:
```diff
@@ -0,0 +1,300 @@
+    This function suppresses the effects of Gibbs oscillations based on the
+    analysis of local total variation (TV). Although artefact correction is
+    done based on two adjanced points for each voxel, total variation should be
+    accessed in a larger range of neighbours. The number of the neighbours to
```
```suggestion
    accessed in a larger range of neighbours. The number of neighbours to
```

arokem commented on dipy/denoise/gibbs.py:
```diff
@@ -0,0 +1,300 @@
+    analysis of local total variation (TV). Although artefact correction is
+    done based on two adjanced points for each voxel, total variation should be
+    accessed in a larger range of neighbours. The number of the neighbours to
+    be considered in TV calculation can be adjusted using parameter n_points.
```
```suggestion
    be considered in TV calculation can be adjusted using the parameter `n_points`.
```

arokem commented on dipy/denoise/gibbs.py:
```diff
@@ -0,0 +1,300 @@
+        Matrix containing one volume (3D) or multiple (4D) volumes of images.
+    slice_axis : int (0, 1, or 2)
+        Data axis corresponding to the number of acquired slices. Default is
+        set to the third axis
```
```suggestion
        set to the third axis.
```

arokem commented on dipy/denoise/tests/test_gibbs.py:
```diff
@@ -0,0 +1,190 @@
+    assert_array_almost_equal(image4d_cor[:, :, 1, 1], image_cor)
+
+
+def test_swaped_gibbs_2d():
```
```suggestion
def test_swapped_gibbs_2d():e
```

arokem commented on dipy/denoise/tests/test_gibbs.py:
```diff
@@ -0,0 +1,190 @@
+    assert_array_almost_equal(image_cor2, image_cor)
+
+
+def test_swaped_gibbs_3d():
```
```suggestion
def test_swapped_gibbs_3d():
```

arokem commented on dipy/denoise/tests/test_gibbs.py:
```diff
@@ -0,0 +1,190 @@
+    assert_array_almost_equal(image3d_cor[1, :, :], image_cor)
+
+
+def test_swaped_gibbs_4d():
```
```suggestion
def test_swapped_gibbs_4d():
```

arokem commented on doc/examples/denoise_gibbs.py:
```diff
@@ -0,0 +1,296 @@
+can be magnified in derived diffusion-based estimates [1]_, [2]_.
+
+In the following example, we show how to suppress Gibbs artefacts of MRI
+images. This algorithm is based on an adapted version of a sub-voxel
```
```suggestion
This algorithm is based on an adapted version of a sub-voxel
```

arokem commented on doc/examples/denoise_gibbs.py:
```diff
@@ -0,0 +1,296 @@
+general, in the context of diffusion-weighted imaging, Gibbs oscillations
+can be magnified in derived diffusion-based estimates [1]_, [2]_.
+
+In the following example, we show how to suppress Gibbs artefacts of MRI
```
```suggestion
In the following example, we show how to suppress Gibbs artefacts of MR images.
```

arokem commented on doc/examples/denoise_gibbs.py:
```diff
@@ -0,0 +1,296 @@
+In the following example, we show how to suppress Gibbs artefacts of MRI
+images. This algorithm is based on an adapted version of a sub-voxel
+Gibbs suppression procedure [3]_. Full details of the implemented algorithm
+can be found in the chapter 3 of [4]_  (please cite [3]_, [4]_ if you are using
```
```suggestion
can be found in chapter 3 of [4]_  (please cite [3]_, [4]_ if you are using
```

arokem commented on doc/examples/denoise_gibbs.py:
```diff
@@ -0,0 +1,296 @@
+can be found in the chapter 3 of [4]_  (please cite [3]_, [4]_ if you are using
+this code).
+
+In this example, we show how to suppress Gibbs of MR images. The algorithm to
```
```suggestion
The algorithm to
```

arokem commented on doc/examples/denoise_gibbs.py:
```diff
@@ -0,0 +1,296 @@
+from dipy.denoise.gibbs import gibbs_removal
+
+""" We first apply this algorithm to T1-weighted dataset which can be fetched
+using the following command lines.
```
```suggestion
using the following code:
```

arokem commented on doc/examples/denoise_gibbs.py:
```diff
@@ -0,0 +1,296 @@
+   in the right panel.
+
+The image artificially corrupted with Gibb's artefacts is shown in the left
+panel. In this panels, the characteristic ringing profile of Gibbs artefacts
```
```suggestion
panel. In this panel, the characteristic ringing profile of Gibbs artefacts
```

arokem commented on doc/examples/denoise_gibbs.py:
```diff
@@ -0,0 +1,296 @@
+
+
+Now let's show how to use the Gibbs suppression algorithm in diffusion-weighted
+images. In this way, we fetch the multi-shell diffusion-weighted dataset which
```
```suggestion
We fetch the multi-shell diffusion-weighted dataset which
```

arokem commented on doc/examples/denoise_gibbs.py:
```diff
@@ -0,0 +1,296 @@
+
+bvals = [200, 400, 1000, 2000]
+
+img, gtab = read_cenir_multib(bvals)
```
I don't think this is the Hansen and Jespersen dataset! That one is `cfin_multib`. This one was provided by Romain Valabregue from CENIR.

arokem commented on doc/examples/denoise_gibbs.py:
```diff
@@ -0,0 +1,296 @@
+
+data_corrected = gibbs_removal(data_slices, slice_axis=2)
+
+""" Due to the high dimensionality of diffusion-weighted data, we recommend you
```
```suggestion
""" Due to the high dimensionality of diffusion-weighted data, we recommend that
```

arokem commented on doc/examples/denoise_gibbs.py:
```diff
@@ -0,0 +1,296 @@
+data_corrected = gibbs_removal(data_slices, slice_axis=2)
+
+""" Due to the high dimensionality of diffusion-weighted data, we recommend you
+to specify which is the axis of data matrix that corresponds to different
```
```suggestion
you specify which is the axis of data matrix that corresponds to different
```

arokem commented on doc/examples/denoise_gibbs.py:
```diff
@@ -0,0 +1,296 @@
+   :align: center
+
+   Uncorrected (left panel) and corrected (middle panel) b-value=0 images. For
+   a reference, the difference between uncorrected and corrected images are
```
```suggestion
   reference, the difference between uncorrected and corrected images is
```

arokem commented on doc/examples/denoise_gibbs.py:
```diff
@@ -0,0 +1,296 @@
+   a reference, the difference between uncorrected and corrected images are
+   shown in the right panel.
+
+The above figure show that the benefits of suppressing Gibbs artefacts is hard
```
```suggestion
The above figure shows that the benefits of suppressing Gibbs artefacts are hard
```

arokem commented on doc/examples/denoise_gibbs.py:
```diff
@@ -0,0 +1,296 @@
+   shown in the right panel.
+
+The above figure show that the benefits of suppressing Gibbs artefacts is hard
+to observed on b-value=0 data. Therefore, diffusion derived metrics for both
```
```suggestion
to observe on b-value=0 data. Therefore, diffusion derived metrics for both
```

skoudoro commented:
Thank you @RafaelNH for addressing @arokem comments!  Any other comment? It will be good if we can merge this PR next week

arokem commented:
I believe this is good to go!

RafaelNH commented:
Done! All suggestion addressed, rebased, and example updated according to latest version of dipy.

arokem commented:
OK. Once the CI goes green, I will merge this.

arokem commented:
Thanks @RafaelNH for the strong work!
